### PR TITLE
Add Rate Limiter utility to akd | Apply to semantic scholar search tool

### PR DESCRIPTION
**Pull request by NISH1001**
# Rate Limiter Implementation for API Requests

## Summary :memo:
This PR introduces a comprehensive rate limiting system to the AKD framework to prevent API quota exhaustion and ensure respectful usage of external services. The implementation includes a thread-safe, async-friendly `RateLimiter` class and its integration with the Semantic Scholar Search Tool.

## Details
_Describe more what you did on changes._
1. **Added RateLimiter Class** (`akd/utils.py`):
   - Thread-safe implementation using `asyncio.Lock`
   - Configurable rate limiting (requests per second)
   - Async-friendly with proper delay handling
   - Comprehensive logging for debugging

2. **Integrated Rate Limiter with Semantic Scholar Search Tool**:
   - Added `requests_per_second` configuration parameter
   - Automatic rate limiter initialization in tool constructor
   - Rate limiting applied before each API request in `_fetch_search_page` method
   - Backward compatible configuration

3. **Comprehensive Test Suite** (`tests/test_rate_limiter.py`):
   - Basic functionality tests (creation, timing accuracy)
   - Configuration testing for various rate limits
   - Concurrency and thread safety tests
   - Integration tests with mock API calls
   - Edge case handling (zero rates, high/low rates)
   - Direct integration testing with SemanticScholarSearchTool

### Basic Rate Limiter Usage
```python
from akd.utils import RateLimiter
import requests

async def make_request():
    rate_limiter = RateLimiter(max_calls_per_second=1.0)
    await rate_limiter.acquire()  # Enforces rate limit
    response = requests.get("https://api.example.com/endpoint")
    return response
```

### Semantic Scholar Tool with Rate Limiting
```python
from akd.tools.search import (
    SemanticScholarSearchTool,
    SemanticScholarSearchToolConfig,
    SemanticScholarSearchToolInputSchema,
)

# Configure tool with rate limiting
config = SemanticScholarSearchToolConfig(
    requests_per_second=0.5,  # 0.5 requests per second (2 second intervals)
)

tool = SemanticScholarSearchTool(config, debug=True)

# Use the tool - rate limiting is automatically applied
result = await tool.arun(
    SemanticScholarSearchToolInputSchema(
        queries=["nepal landslide", "nepal earthquake"],
        max_results=50,
    )
)
```

### Key Features
- **Configurable Rate Limits**: Set `requests_per_second` in tool configuration
- **Automatic Rate Limiting**: Applied transparently before each API request  
- **Thread-Safe**: Safe for concurrent usage across multiple async tasks
- **Debug Logging**: Detailed logs when rate limiting delays occur
- **Backward Compatible**: Existing code continues to work without changes

## Checks
- [x] Added comprehensive pytest test suite with 65%+ coverage
- [x] Tested rate limiting accuracy and timing behavior
- [x] Verified thread safety with concurrent requests
- [x] Tested integration with SemanticScholarSearchTool
- [x] All existing tests continue to pass
- [x] Rate limiting respects API provider guidelines
- [x] Backward compatibility maintained

### CODE REVIEW

File: akd/tools/search/semantic_scholar_search.py [hunk 1]
```diff
@@ -171,6 +194,9 @@ async def _fetch_search_page(
                 logger.debug("Using API Key.")
 
         try:
+            # Apply rate limiting before making the request
+            await self.rate_limiter.acquire()
```
Comment (NISH1001) [inline akd/tools/search/semantic_scholar_search.py:198]: See the usage section to understand this. Basically, we first acquire the lock to make ready for the requests (both async and sync).

**Review by NISH1001 — COMMENTED** [review]
One comment on acquiring the lock

**Review by movinam — APPROVED** [review]
Looks good. Thanks!
